### PR TITLE
[SERVICE-130] SnapGroup Validation

### PR DESCRIPTION
**Pull request by tomrobertsOF**
Added implementation for validate group function. Will now split the group if it becomes disjoint.

### CODE REVIEW

**Comment by tomrobertsOF** [comment]
@MichaelMCoates I've got what I think should be a fix in place for the tabbing issue. 

Having talked it over a bit, there may have some unintended side-effects on S&R. I think we should leave this open until that PR is merged so we can test more thoroughly before merging this one in.

File: src/provider/snapanddock/SnapService.ts [hunk 1]
```diff
@@ -235,7 +235,8 @@ export class SnapService {
         for (let i = 0; i < windows.length; i++) {
             adjacencyList[i] = [];
             for (let j = 0; j < windows.length; j++) {
-                if (i !== j && RectUtils.distance(windows[i].getState(), windows[j].getState()).border(0)) {
+                const distance = RectUtils.distance(windows[i].getState(), windows[j].getState());
```
Comment (pjbroadbent) [inline src/provider/snapanddock/SnapService.ts]: Unused var `distance`?

File: src/provider/snapanddock/SnapService.ts [hunk 2]
```diff
@@ -264,5 +265,22 @@ export class SnapService {
...
             }
         }
+
+        function isAdjacent(win1: Snappable, win2: Snappable) {
+            const distance = RectUtils.distance(win1.getState(), win2.getState());
+            if (win1.getTabGroup() && win1.getTabGroup() === win2.getTabGroup()) {
+                // Special handling for tab groups. When validating, all windows in a tabgroup are
+                // assumed to be adjacent to avoid weirdness with hidden windows.
+                return true;
+            } else if (win1.getState().hidden || win1.getState().hidden) {
```
Comment (pjbroadbent) [inline src/provider/snapanddock/SnapService.ts]: Should one of these be `win2`?..

Comment (tomrobertsOF) [reply src/provider/snapanddock/SnapService.ts]: Yes

File: src/provider/snapanddock/SnapService.ts (hunk 1 above)
Comment (tomrobertsOF) [reply src/provider/snapanddock/SnapService.ts]: Yea, got moved to the util and forgot to remove this.

**Pull request merged**